lib/TH/THGeneral.c

```diff
@@ -139,17 +142,23 @@ static long getAllocSize(void *ptr) {
  *     soft max by 40%
  */
 static void maybeTriggerGC() {
-  if(torchGCFunction && torchHeapSize > torchHeapSizeSoftMax) {
+  if(torchGCFunction && heapSize > heapSoftmax ) {
     torchGCFunction(torchGCData);
-    if(torchHeapSize > torchHeapSizeSoftMax * 0.8) {
-      torchHeapSizeSoftMax = torchHeapSizeSoftMax * 1.4;
+    if(heapSize > heapSoftmax * heapSoftmaxGrowthThresh) {
+      heapSoftmax = heapSoftmax * heapSoftmaxGrowthFactor;
     }
   }
 }
 
 // hooks into the TH heap tracking
 void THHeapUpdate(long size) {
-  torchHeapSize += size;
+  THAtomicAddLong(&heapSize, size);
+
+# ifdef TH_CHECK_HEAP_UPDATE
+  if (heapSize < 0)
```

Comment:
I think we need to use atomics to _read_ the heap size too. Otherwise there's no guarantee that an atomic write by another thread is seen by this thread.
You could also use the return value from THAtomicAddLong and work with that instead of the global `heapSize`, if you're not concerned about concurrent updates by other threads during `THHeapUpdate`.